transport/grpc/outbound_test.go

```diff
@@ -200,3 +203,104 @@ func TestCallStreamWithInvalidPeer(t *testing.T) {
 		}.Error(),
 	)
 }
+
+func TestCallSvcMatch(t *testing.T) {
+	const (
+		successType  = "success"
+		failType     = "fail"
+		noheaderType = "noheader"
+	)
+	tests := []struct {
+		msg        string
+		serverPort string
+		testType   string
+		wantErr    bool
+	}{
+		{
+			"test call service match success",
+			"15213",
+			successType,
+			false,
+		},
+		{
+			"test call service match failed",
+			"15619",
+			failType,
+			true,
+		},
+		{
+			"test no service name response header",
+			"15640",
+			noheaderType,
+			false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			server := grpc.NewServer(
+				grpc.CustomCodec(customCodec{}),
+				grpc.UnknownServiceHandler(func(srv interface{}, stream grpc.ServerStream) error {
+					ctx := stream.Context()
+					// Get transport request
+					md, ok := metadata.FromIncomingContext(ctx)
+					if md == nil || !ok {
+						return yarpcerrors.Newf(yarpcerrors.CodeInternal, "cannot get metadata from ctx: %v", ctx)
+					}
+					transportRequest, err := metadataToTransportRequest(md)
+					if err != nil {
+						return err
+					}
+
+					responseWriter := newResponseWriter()
+					defer responseWriter.Close()
+
+					if tt.testType == successType {
+						// Echo accepted rpc-service in response header
+						responseWriter.AddSystemHeader(ServiceHeader, transportRequest.Service)
+					} else if tt.testType == failType {
+						// Append a wrong service name
+						responseWriter.AddSystemHeader(ServiceHeader, "ThisIsAWrongSvcName")
+					}
+
+					// Send the response attributes back and end the stream.
+					if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+						// We couldn't send the response.
+						return sendErr
+					}
+					if responseWriter.md != nil {
+						stream.SetTrailer(responseWriter.md)
+					}
+					return err
+				}),
+			)
+			listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%s", tt.serverPort))
+			require.NoError(t, err)
+			go func() {
+				err := server.Serve(listener)
+				require.NoError(t, err)
+			}()
+			defer server.Stop()
+
+			grpcTransport := NewTransport()
+			out := grpcTransport.NewSingleOutbound(listener.Addr().String())
+			require.NoError(t, grpcTransport.Start())
+			require.NoError(t, out.Start())
+			defer grpcTransport.Stop()
+			defer out.Stop()
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			req := &transport.Request{
+				Service:   "Service",
+				Procedure: "Hello",
+				Body:      bytes.NewReader([]byte("world")),
```

Comment:
Is this necessary for running the test? Let's try and have the this test do as few things as possible to make it easier for others with less/no context to review.

Reply:
Yes, "Service" field is what we need, "Procedure" and "Body" are necessary to avoid empty body or procedure exception from server side.